assignments/burmanm/week-3.scm

```diff
@@ -0,0 +1,279 @@
+(define (integral f a b n)
+  (* (/ (h a b n) 3) (integ f a b n)))
+
+;; Exercise 1.30
+(define (inc n) (+ n 1))
+
+(define (sum term a next b)
+  (define (iter a result)
+    (if (> a b)
+	result
+	(iter (next a) (+ result (term a)))))
+  (iter a 0))
+
+;; Exercise 1.31
+
+(define (product term a next b)
+  (define (iter a result)
+    (if (> a b)
+	result
+	(iter (next a) (* result (term a)))))
+  (iter a 1))
+
+(define (product-recursive term a next b)
+  (if (> a b)
+      1
+      (+ (term a)
+	 (product-recursive term (next a) next b))))
+
+;; define factorial
+;; define approx-pi
+
+;; (define approx-pi 
+;; (product 
+
+;; Exercise 1.32
+
+;; (accumulate combiner null-value term a next b)
+(define (accumulate-recursive combiner null-value term a next b)
+  (if (> a b)
+      null-value
+      (combiner (term a)
+	 (accumulate-recursive combiner null-value term (next a) next b))))
+
+(define (accumulate combiner null-value term a next b)
+  (define (iter a result)
+    (if (> a b)
+	result
+	(iter (next a) (combiner result (term a)))))
+  (iter a null-value))
+
+(define (sum term a next b)
+  (accumulate + 0 term a next b))
+
+(define (product term a next b)
+  (accumulate * 1 term a next b))
+
+;; Exercise 1.33
+
+(define (filtered-accumulate filter combiner null-value term a next b)
+  (define (iter a result filter)
+    (if (> a b)
+	result
+	(iter (next a) (combiner (if (filter a) (term a) null-value)))))
+  (iter a null-value filter))
+
+(define (sum-square-prime a b)
+  (filtered-accumulate prime? + 0 square a inc b))
+
+(define (coprime? x) (= (gcd x n) 0))
+(define (sum-coprime n)
+  (filtered-accumulate coprime? * 1 identity 1 inc n))
+
+;; Exercise 1.34
+
+;; scheme@(guile-user)>(define (f g) (g 2))
+;; scheme@(guile-user)> ,trace (f f)
+;; trace: (f #<procedure f (g)>)
+;; trace: (f 2)
+;; trace: (_ 2)
+;; trace: (throw wrong-type-arg #f "Wrong type to apply: ~S" (2) (2))
+
+
+;; We end up with evaluation (2 2) which can't be applied as 2 isn't a defined function
+
+;; Exercise 1.35
+
+;; x^2 = x + 1
+;; Solving to x gives us x = (1 +- sqrt(5))/2
+(fixed-point (lambda (x) (+ 1 (/ 1 x))) 1.0)
+
+;;]=> (fixed-point (lambda (x) (+ 1 (/ 1 x))) 1.0)
+;Value: 1.6180327868852458
+;; Which is close enough..
+
+;; Exercise 1.36
+
+(define (fixed-point f first-guess)
+  (define (close-enough? v1 v2)
+    (< (abs (- v1 v2))
+       tolerance))
+  (define (try guess)
+    (let ((next (f guess)))
+      (newline)
+      (display guess)
+      (if (close-enough? guess next)
+	  next
+	  (try next))))
+  (try first-guess))
+
+(fixed-point (lambda (x) (/ (log 1000) (log x))) 2.0)
+
+;; Takes 34 guesses to reach 4.555532270803653
+
+;; With average-damping:
+
+(define (average x y) (/ (+ x y) 2))
+(fixed-point (lambda (x) (average x (/ (log 1000) (log x)))) 2.0)
+
+;; We get 9 iterations to reach 4.555537551999825
+;; If we select different initial guess, then it may take different amount of iterations (such as 10.0 requiring 10 iterations)
+
```

Comment:
I measured 32 and 8 iterations with a tolerance of 0.00001. What value did you used?